### PR TITLE
Make DSP time constants frame-rate independent

### LINKED ISSUE #152: Make DSP time constants frame-rate independent
## Motivation
With user-configurable STFT frame sizes (23-93 ms in noise-repellent), engine behavior audibly changes with frame size beyond the intended time/frequency-resolution tradeoff: larger frames sound increasingly robotic/smeared on transient material.

## Cause
A class of constants is counted in frames/hops rather than seconds, so real-time spans scale with frame size (hops at 48 kHz: 276 samples at 23 ms vs 1116 at 93 ms):

- NLM patch (8 frames): 46 ms -> 186 ms temporal smear; full NLM context ~167 ms -> ~674 ms
- Gain attack/release envelopes quantize to 23 ms hops (sub-hop attacks step, musical chatter); backward masking (10 ms tau) effectively dies
- Per-frame IIR alphas (Martin/SPP/transient/masking smoothers) stretch 4x in effective tau
- Martin minimum-statistics window: ~0.5 s -> ~2.2 s; median updates every ~186 ms
- Smoothing-mode crossfade (30 ms) collapses to ~1 frame

Measured on a harmonic+noise probe (learn-then-denoise @48 kHz): time fidelity degrades monotonically 23 -> 93 ms while spectral match slightly improves - the expected tradeoff curve, amplified by the above.

## Scope
Normalize time constants to seconds: frame-rate-aware NLM patch sizing (or fixed-ms patch), per-second IIR alphas, hop-count-derived transition lengths, ms-based estimator windows. Retune risk: adaptive estimator behavior was tuned around ~46 ms frame rates, so this needs A/B quality validation, not just a mechanical conversion.

## Acceptance
- Same perceptual quality at 23 ms and 93 ms on transient material (or quantified tradeoff with no time-smearing beyond hop granularity)
- Existing test suite green plus a frame-size-invariance regression test

### PR DESCRIPTION
**Pull request by lucianodato**
Closes #152.

## Motivation
With user-configurable STFT frame sizes (23–93ms in noise-repellent), engine behavior audibly changed with frame size beyond the intended time/frequency-resolution tradeoff: larger frames sounded increasingly robotic/smeared on transient material.

## Cause
Time constants counted in frames/hops rather than seconds, so real-time spans scaled with frame size (hops at 48kHz: 276 samples at 23ms vs 1116 at 93ms).

## What changed
- **NLM fixed-ms geometry** (`src/shared/frame_rate_norm.h`, new): patch 92ms (Lukin AES123-anchored: 8 frames at ~11.5ms hop), **symmetric 128ms past / 128ms future** search (restoration-oriented choice — latency is not a product constraint), recomputed per frame size with a frozen per-option table (23ms→16/22/22, 32→12/16/16, 46→8/11/11, 64→6/8/8, 93→4/6/6) plus formula fallback, and Hz-anchored frequency axis (~170Hz search/paste).
- **Per-hop tau retuning** (init-time `set_hop_sec()`, legacy 12.5ms-hop behavior preserved): Martin/SPP/transient/HPSS/veto smoothing alphas.
- **Ms-based estimator windows**: Martin subwindow 138ms, median window 287.5ms / update 92ms (buffer grown to 64 max), Brandt stats interval 46ms.
- **Hop bug fixes**: Brandt history factor `0.5→/OVERLAP_FACTOR`; masking estimator and gain smoother now use the true STFT hop (frame/overlap, not FFT/overlap) via `spectral_denoiser_initialize_with_hop()` (old init kept as wrapper).
- **Crossfade floor**: smoothing-mode transition is `max(30ms-in-frames, 4 frames)`.
- **NLM halo 4→8 frames** (covers patch ≤16) and **chunked-SIMD patch distance** (8/4/1) for non-8 patch sizes — 23ms file render 9.9s→2.7s on M4 (~3.7x); patch-8 path kept bit-identical.

## Latency impact (reported to host, ms is SR-independent)
23ms: 46→69 · 32ms: 64→80 · 46ms: 92 (unchanged) · 64ms: 128→112 · 93ms: 186→139.5.

## Validation
- New `test_frame_size_invariance` (transient material, NLM, 23 vs 93ms, latency-aligned): normalized pre-echo/tail smear ratios must match within 1.6x. Verified it **fails on main** (2.2x) and **passes here** (1.2x).
- Full suite green: 33/33 (`-DENABLE_TESTS=ON`). Reference wavs regenerated via the documented workflow (DSP output intentionally changed); the two `test_specbleach_{stereo,cpp_smoke}` link failures seen mid-work were stale binaries, gone after a clean rebuild.
- Remaining: listening A/B on transient material before merge (operating point moved off the legacy tuning).

## Summary by CodeRabbit

- **New Features**
  - Added configurable STFT hop-size support for denoising.
  - Improved consistency of noise estimation, smoothing, transient detection, masking, and suppression across different frame sizes.
  - Added frame-rate-aware NLM filtering geometry and expanded support for variable patch sizes.

- **Bug Fixes**
  - Reduced frame-size-dependent temporal smearing and corrected latency handling.
  - Added safeguards for invalid or zero hop durations.

- **Tests**
  - Added coverage for frame-rate normalization, hop-size configuration, edge cases, and frame-size invariance.

### CODE REVIEW

File: src/processors/denoiser/spectral_denoiser.c [hunk 1]
```diff
@@ -271,16 +280,22 @@ SpectralProcessorHandle spectral_denoiser_initialize(
     return NULL;
   }
 
+  // Frame-rate normalization: fixed-ms / fixed-Hz geometry + per-hop alphas.
+  const float hop_sec = self->hop_sec;
+  const float frame_ms = hop_sec * 4000.0F; // hop = frame/4
```
[the remark this answers is not in the document]
Comment (lucianodato) [reply src/processors/denoiser/spectral_denoiser.c]: Fixed — frame duration now derives from overlap_factor (exact with explicit hop; fft-derived on the legacy path). Thanks for catching it.

File: src/shared/denoiser_logic/estimators/brandt_noise_estimator.c [hunk 2]
```diff
@@ -68,13 +70,14 @@ BrandtNoiseEstimator* brandt_noise_estimator_initialize(
...
+  // Calculate history size from duration. hop == frame/OVERLAP_FACTOR;
+  // the old *0.5 (50% overlap) assumption is fixed to /OVERLAP_FACTOR here.
   float ms_per_frame = (float)fft_size * 1000.0f / (float)sample_rate;
   // Overlap consideration: Usually frame step is hop_size.
   // Assuming hop = fft_size/2 or similar? The caller usually provides
   // parameters. If exact duration needed, we might need hop_size.
-  // Assuming standard 50% overlap for calculation roughly:
-  float frame_duration = ms_per_frame * 0.5f; // Rough approximation of step
+  // Fixed 4x-overlap conversion (was 50% approximation):
+  float frame_duration = ms_per_frame / (float)OVERLAP_FACTOR;
```
[the remark this answers is not in the document]
Comment (lucianodato) [reply src/shared/denoiser_logic/estimators/brandt_noise_estimator.c:82]: Fixed — set_hop_sec now also rebuilds the history buffer for the true hop (init-time realloc, keeps existing storage on failure), instead of only retuning the stats interval.

File: src/shared/denoiser_logic/processing/masking_veto.c [hunk 3]
```diff
@@ -288,3 +291,12 @@ void masking_veto_apply(MaskingVeto* self, const float* smoothed_spectrum,
     alpha[k] = fmaxf(ALPHA_MIN, alpha[k] * (1.0F - veto_amount));
   }
 }
+
+void masking_veto_set_hop_sec(MaskingVeto* self, float hop_sec) {
+  if (!self || !(hop_sec > 0.0F)) {
+    return;
+  }
+  self->smooth =
+      sb_alpha_retuned(MASKING_VETO_SMOOTHING, LEGACY_REF_HOP_SEC, hop_sec);
```
[the remark this answers is not in the document]
Comment (lucianodato) [reply src/shared/denoiser_logic/processing/masking_veto.c:302]: Fixed — you were right, the weight was inverted: alpha is the memory coefficient everywhere else, so the clean-signal update now reads (1-smooth)*current + smooth*stable. Identical output at the reference hop (0.5). Good catch.

File: src/shared/frame_rate_norm.h [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    g.past = 6U;
+    g.future = 6U;
+  } else if (hop_sec > 0.0F) {
+    g.patch = sb_frames_for_ms(NLM_PATCH_TIME_MS, hop_sec, 4U, 16U);
+    g.past = sb_frames_for_ms(NLM_SEARCH_PAST_MS, hop_sec, 8U, 32U);
+    g.future = sb_frames_for_ms(NLM_SEARCH_FUTURE_MS, hop_sec, 8U, 32U);
+  }
+  if (bin_hz > 0.0F) {
+    g.search_freq = sb_bins_for_hz(NLM_SEARCH_FREQ_HZ, bin_hz, 2U, 32U);
+    g.paste = sb_bins_for_hz(NLM_PASTE_FREQ_HZ, bin_hz, 2U, 16U);
```
[the remark this answers is not in the document]
Comment (lucianodato) [reply src/shared/frame_rate_norm.h:153]: Skipping — the bounds are call-site-specific (4/16 patch, 8/32 past, 2/32 search, 2/16 paste, 1/16 stats) and naming each would add indirection without meaning. The NLM patch/halo limits did move to configurations.h as shared constants.

File: src/shared/utils/masking_estimator.c [hunk 5]
```diff
@@ -169,6 +172,20 @@ MaskingEstimator* masking_estimation_initialize(
...
+    return;
+  }
+  for (uint32_t j = 0U; j < self->number_critical_bands; j++) {
+    const float bark = fminf((float)j, 24.0F);
+    const float weight = bark / 24.0F;
+    const float tau = ((1.0F - weight) * FORWARD_MASKING_TAU_LOW_MS) +
+                      (weight * FORWARD_MASKING_TAU_HIGH_MS);
+    self->forward_decays[j] = expf(-hop_sec / tau);
+  }
+  self->backward_decay = expf(-hop_sec / BACKWARD_MASKING_TAU_MS);
```
[the remark this answers is not in the document]
Comment (lucianodato) [reply src/shared/utils/masking_estimator.c]: Not applied — and it must not be: the constants hold seconds despite the _MS names (0.100F = 100ms), so expf(-hop_sec/tau) is already correct and the suggested change would make masking ~1000x too fast. Renamed them to _SEC instead to disarm the trap. The misleading suffix was the real issue.

**Comment by lucianodato** [comment]
@coderabbitai review

All review findings from the last round are addressed in the latest commits — please re-review and approve if everything looks good.

**Fixed:**
- Frame duration now derives from `overlap_factor` instead of hardcoded /4.
- Brandt history storage rebuilds for the true hop in `set_hop_sec` (init-time realloc, keeps existing storage on allocation failure).
- Masking-veto stabilization weight was genuinely inverted vs the alpha-as-memory convention — flipped the clean-signal update (bit-identical at the 0.5 reference hop).
- NLM patch/halo limits moved into `configurations.h` as shared constants.
- Smoother test now asserts against the explicitly configured true hop; masking-estimator test asserts behavioral difference across hops; invariance test uses a unified peak window.

**Deliberately not fixed:**
- *Named clamp bounds in `frame_rate_norm.h`*: the bounds are call-site-specific (4/16 patch, 8/32 past, 2/32 search, 2/16 paste, 1/16 stats) — naming each would add indirection without meaning.
- *Masking ms→sec conversion*: not applied on purpose. The constants hold **seconds** despite the legacy `_MS` names (0.100F = 100ms), so the suggested change would have made masking ~1000x too fast. Renamed them to `_SEC` instead to disarm the trap.

Full suite green locally (34/34) and all CI checks pass.

**Comment by lucianodato** [comment]
@coderabbitai approve
